### ISSUE TITLE
Spring-Batch not supporting multi-tenancy with database backed job repository

### ISSUE THREAD

**Issue by itsmedtk** [spring-projects/spring-batch#3675]
Scenario/Issue: I have a multi-tenant environment with each tenant having a separate schema on Oracle database. Every time a spring batch job is started, it persists the job related data in the respective schema of the tenant.

But, when I start multiple jobs on 2 or more tenants in parallel, and if the jobIds of different tenants happen to get the same id based on the respective schema's job sequence, then I am running to concurrency issues (see below) as the Entity.java class has equals method on 'id' only.

This seems to be a limitation/bug as Spring Batch expects the Id to be different/unqiue for every job, which may not be true for database backed JobRepository when multiple tenant jobs are running in parallel.

"java.util.ConcurrentModificationException
	at java.util.ArrayList$Itr.checkForComodification(Unknown Source)
	at java.util.ArrayList$Itr.next(Unknown Source)
	at com.XXXX.inference.springbatch.InferenceReader.open(InferenceReader.java:353)
	at sun.reflect.GeneratedMethodAccessor3206.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:197)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.support.DelegatingIntroductionInterceptor.doProceed(DelegatingIntroductionInterceptor.java:136)
	at org.springframework.aop.support.DelegatingIntroductionInterceptor.invoke(Delega..."

Possible Solution: I fixed the issue by adding ThreadLocal to be able to set tenant, and updating equals() in both Entity.java and StepExuection.java. See attached files for the changes (we used spring-batch-core-4.2.1.RELEASE). You may fix in a different way, but this I think should be fixed in the code.

**Comment by itsmedtk** [comment]
[spring-batch-mt-fix.zip](https://github.com/spring-projects/spring-batch/files/4268185/spring-batch-mt-fix.zip)

**Comment by benas** [comment]
This is a duplicate of #1122.
